sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java

```diff
@@ -0,0 +1,149 @@
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.reference.sys.shard;
+
+import com.google.common.base.Suppliers;
+import io.crate.blob.v2.BlobShard;
+import io.crate.metadata.IndexParts;
+import io.crate.metadata.PartitionName;
+import io.crate.metadata.RelationName;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.lucene.BytesRefs;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.StoreStats;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class ShardRowContext {
+
+    private static final BytesRef EMPTY_BYTES_REF = new BytesRef("");
+
+    private final IndexShard indexShard;
+    @Nullable
+    private final BlobShard blobShard;
+    private final ClusterService clusterService;
+    private final Supplier<Long> sizeSupplier;
+    private final IndexParts indexParts;
+    private final BytesRef partitionIdent;
+    private final int id;
+    private final BytesRef path;
+    @Nullable
+    private final BytesRef blobPath;
+    @Nullable
+    private final String aliasName;
+    @Nullable
+    private final String templateName;
+
+    public ShardRowContext(IndexShard indexShard, ClusterService clusterService) {
+        this(indexShard, null, clusterService, Suppliers.memoizeWithExpiration(() -> {
+            StoreStats storeStats = indexShard.storeStats();
+            if (storeStats == null) {
+                // will return null if already closed due to shard deletion
+                return 0L;
+            }
+            return storeStats.getSizeInBytes();
+        }, 10, TimeUnit.SECONDS));
+    }
+
+    public ShardRowContext(BlobShard blobShard, ClusterService clusterService) {
+        this(blobShard.indexShard(), blobShard, clusterService, blobShard::getTotalSize);
+    }
+
+    private ShardRowContext(IndexShard indexShard,
+                            @Nullable BlobShard blobShard,
+                            ClusterService clusterService,
+                            Supplier<Long> sizeSupplier) {
```

Comment:
what size is this? also is this used? didn't find any usage of the `size` method

Reply:
see https://github.com/crate/crate/pull/7594/files#diff-5a4815a0678a6f35b2fa4aaa3aa8742eR141